### PR TITLE
HAWQ-1404. PXF to leverage file-level stats of ORC file and emit records for COUNT(*).

### PR DESCRIPTION
**Pull request by sansanichfb**

### CODE REVIEW

**Comment by shivzone** [comment]
Do you plan to add any unit tests ?

File: pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/utilities/Utilities.java [hunk 1]
```diff
@@ -151,4 +157,78 @@ public static String maskNonPrintables(String input) {
...
+     * @return fragment metadata
+     * @throws IllegalArgumentException if fragment metadata information wasn't found in input data
+     * @throws Exception
+     */
+    public static FragmentMetadata parseFragmentMetadata(InputData inputData) throws Exception {
+        byte[] serializedLocation = inputData.getFragmentMetadata();
+        if (serializedLocation == null) {
+            throw new IllegalArgumentException("Missing fragment location information");
+        }
+        try (ByteArrayInputStream bytesStream = new ByteArrayInputStream(serializedLocation);
```
Comment (denalex) [inline pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/utilities/Utilities.java]: since you're not using bytesStream directly, can collapse to 
ObjectInputStream objectStream = new ObjectInputStream( new ByteArrayInputStream(serializedLocation))

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveORCAccessor.java [hunk 2]
```diff
@@ -65,13 +78,24 @@ public HiveORCAccessor(InputData input) throws Exception {
...
-        addFilters();
-        return super.openForRead();
+        if (useStats) {
+            orcReader = HiveUtilities.getOrcReader(inputData);
+            if (orcReader == null) {
+                return false;
+            }
+            retrieveStats();
+            objectsEmitted = 0;
+            return super.openForRead();
```
Comment (denalex) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveORCAccessor.java]: should carry this common line outside if / else ?

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveORCAccessor.java [hunk 3]
```diff
@@ -213,4 +237,40 @@ private boolean buildArgument(SearchArgument.Builder builder, Object filterObj)
...
+     * Emits tuple without reading from disk, currently supports COUNT
+     */
+    @Override
+    public OneRow emitAggObject() {
+        OneRow row = null;
+        switch (inputData.getAggType()) {
+            case COUNT:
+                if (objectsEmitted < count) {
+                    objectsEmitted++;
+                    row = new OneRow(key, data);
```
Comment (denalex) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveORCAccessor.java]: as this is likely a dummy object, can you reuse a single object and return it multiple times rather than creating multiple dummy objects ?

File: pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/AggBridge.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+                    resolvedFieldsCache.put(onerow.getKey(), resolvedFields);
+                }
+                outputQueue = outputBuilder.makeOutput(resolvedFields);
+                if (!outputQueue.isEmpty()) {
+                    output = outputQueue.pop();
+                    break;
+                }
+            }
+        } catch (Exception ex) {
+            throw ex;
```
Comment (denalex) [inline pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/AggBridge.java]: what's the point of this ?

File: pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/utilities/EnumAggregationType.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.hawq.pxf.api.utilities;
+
+public enum EnumAggregationType {
+
+    COUNT("count", true),
+    UNKNOWN("unknown", false);
```
Comment (denalex) [inline pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/utilities/EnumAggregationType.java]: why is this needed, rather than returning null when returning enum lookup from unknown value ?

File: pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/utilities/Utilities.java (hunk 1 above)
Comment (sansanichfb) [inline pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/utilities/Utilities.java]: Thanks, makes sense.

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveORCAccessor.java (hunk 2 above)
Comment (sansanichfb) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveORCAccessor.java]: Sure, updated.

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveORCAccessor.java (hunk 3 above)
Comment (sansanichfb) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveORCAccessor.java]: Good idea, updated.

File: pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/AggBridge.java (hunk 4 above)
Comment (sansanichfb) [inline pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/AggBridge.java]: Added logging of message before re-throwing.

File: pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/utilities/EnumAggregationType.java (hunk 5 above)
Comment (sansanichfb) [inline pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/utilities/EnumAggregationType.java]: Deleted.

**Comment by sansanichfb** [comment]
@shivzone sure, added unit-tests.
